pkg/winperfcounters/watcher.go

```diff
@@ -0,0 +1,147 @@
+
+//go:build windows
+// +build windows
+
+package winperfcounters // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/winperfcounters"
+
+import (
+	"fmt"
+
+	"go.uber.org/multierr"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/winperfcounters/internal/pdh"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/winperfcounters/internal/third_party/telegraf/win_perf_counters"
+)
+
+var _ PerfCounterWatcher = (*Watcher)(nil)
+
+// PerfCounterWatcher represents how to scrape data
+type PerfCounterWatcher interface {
+	// Path returns the counter path
+	Path() string
+	// ScrapeData collects a measurement and returns the value(s).
+	ScrapeData() ([]win_perf_counters.CounterValue, error)
+	// Close all counters/handles related to the query and free all associated memory.
+	Close() error
+
+	GetMetricRep() MetricRep
+}
+
+const instanceLabelName = "instance"
+
+type Watcher struct {
+	Counter *pdh.PerfCounter
+	MetricRep
+}
+
+func (w Watcher) Path() string {
+	return w.Counter.Path()
+}
+
+func (w Watcher) ScrapeData() ([]win_perf_counters.CounterValue, error) {
+	return w.Counter.ScrapeData()
+}
+
+func (w Watcher) Close() error {
+	return w.Counter.Close()
+}
+
+func (w Watcher) GetMetricRep() MetricRep {
+	return w.MetricRep
+}
+
+// BuildPaths creates watchers and their paths from configs.
+func BuildPaths(objs []ObjectConfig) ([]PerfCounterWatcher, error) {
+	var errs error
+	var watchers []PerfCounterWatcher
+
+	for _, objCfg := range objs {
+		for _, instance := range objCfg.instances() {
+			for _, counterCfg := range objCfg.Counters {
+				counterPath := counterPath(objCfg.Object, instance, counterCfg.Name)
+
+				c, err := pdh.NewPerfCounter(counterPath, true)
+				if err != nil {
+					errs = multierr.Append(errs, fmt.Errorf("counter %v: %w", counterPath, err))
+				} else {
+					newWatcher := Watcher{Counter: c}
+
+					if counterCfg.MetricRep.Name != "" {
+						metricCfg := MetricRep{Name: counterCfg.MetricRep.Name}
+						if counterCfg.Attributes != nil {
+							metricCfg.Attributes = counterCfg.Attributes
+						}
+						newWatcher.MetricRep = metricCfg
+					} else {
+						newWatcher.MetricRep.Name = c.Path()
+					}
+
+					watchers = append(watchers, newWatcher)
+				}
+			}
+		}
+	}
+
+	return watchers, errs
+}
+
+func counterPath(object, instance, counterName string) string {
+	if instance != "" {
+		instance = fmt.Sprintf("(%s)", instance)
+	}
+
+	return fmt.Sprintf("\\%s%s\\%s", object, instance, counterName)
+}
+
+type CounterValue struct {
+	MetricRep
+	Value int64
+}
+
+// WatchCounters pulls values given the passed in watchers
+func WatchCounters(watchers []PerfCounterWatcher) (metrics []CounterValue, errs error) {
+	for _, watcher := range watchers {
+		counterValues, err := watcher.ScrapeData()
+		if err != nil {
+			errs = multierr.Append(errs, err)
+			continue
+		}
+		metric := watcher.GetMetricRep()
+
+		if counterValues[0].InstanceName != "" {
```

Comment:
Even though there might be a good reason to assume `counterValues[0]` exists, we should check it. Since we access it several times, might as well make a variable out of it too.

Reply:
Updated